packages/data-context/src/codegen/code-generator.ts

```diff
@@ -0,0 +1,156 @@
+import * as fs from 'fs-extra'
+import { isBinaryFile } from 'isbinaryfile'
+import * as path from 'path'
+import * as ejs from 'ejs'
+import fm from 'front-matter'
+
+export interface Action {
+  templateDir: string
+  target: string
+  overwrite?: boolean
+}
+
+export interface CodeGenResult {
+  status: 'add' | 'overwrite' | 'skipped'
+  type: 'text' | 'binary'
+  file: string
+  content: string
+}
+
+export interface CodeGenResults {
+  files: Array<CodeGenResult>
+  failed: Array<Error>
+}
+
+/**
+ * Utility for generating files from ejs templates or for general scaffolding purposes.
+ * Given a templte directory, all files within will be moved to the target directory specified whilst
+ * maintaining the folder heirarchy. It supports both text and binary files, with text files having the
+ * additional ablity to be rendered with .ejs support meaning any arguments passed in can be interpolated
+ * into the file. For custom file naming, front-matter can be used to specify the output fileName.
+ */
+export async function codeGenerator (
+  action: Action,
+  args: { [key: string]: any },
+): Promise<CodeGenResults> {
+  const templateFiles = await allFilesInDir(action.templateDir)
+  const codeGenResults: CodeGenResults = { files: [], failed: [] }
+
+  for (const file of templateFiles) {
+    const isBinary = await isBinaryFile(file)
+    const parsedFile = path.parse(file)
+
+    const processBinaryFile = async () => {
```

Comment:
What is the use case for this? Are we ever generated a binary file? What kind of file?

Reply:
pngs

Reply:
I took a lot of inspiration from [node-plop](https://github.com/plopjs/node-plop/blob/22aa15e31026d48ddb9a29419597b8fec2d78b02/src/actions/_common-action-add-file.js#L37) and [nx](https://github.com/nrwl/nx/blob/27cc2d9c5f1e90ee4246a314c13b6c87fb7ac026/packages/devkit/src/generators/generate-files.ts#L6). They both had special handling for binary files and it wasn't too difficult to include it so I included it! Might be useful in the future.